File: android/app/src/main/java/com/titanarq/studentassistant/tutor/VoiceQuestion.kt
package com.titanarq.studentassistant.tutor

import com.titanarq.studentassistant.capture.RecognizerEngine
import com.titanarq.studentassistant.capture.RecognizerError
import com.titanarq.studentassistant.capture.RecognizerListener

/** Why a spoken question gave no text. */
enum class VoiceProblem {
    /** Silence, or nothing recognisable. */
    NO_SPEECH,

    /** The microphone permission is missing. */
    PERMISSION_DENIED,

    /** No recognizer on the phone, or none for Spanish. */
    UNAVAILABLE,

    /** Network, server, audio or a busy recognizer. */
    FAILED,
}

/**
 * One spoken question (#248): a single listening round of a [RecognizerEngine] (Android's
 * `SpeechRecognizer` in `es-ES`, the capture screen's engine) with its interim text, which ends when
 * the student stops talking and hands over the recognised text. [stop] ends it early and keeps what
 * was heard so far. Unlike the capture screen's transcriber it never restarts and never talks to a
 * session.
 *
 * Not thread-safe: every call and the engine's callbacks come from the main thread.
 */
class VoiceQuestion(
    private val engine: RecognizerEngine,
    private val language: String = LANGUAGE,
) {
    /** What a listening round reports; exactly one of [onFinal] / [onProblem] ends it. */
    interface Listener {
        /** The question so far, while the student speaks. */
        fun onInterim(text: String)

        /** The recognised question (never blank). */
        fun onFinal(text: String)

        fun onProblem(problem: VoiceProblem)
    }

    /** True from [start] until the round ends. */
    var listening: Boolean = false
        private set

    private var round = 0
    private var heard = ""
    private var listener: Listener? = null

    /** Starts listening for one question; ignored while a round is running. */
    fun start(listener: Listener) {
        if (listening) return
        listening = true
        heard = ""
        this.listener = listener
        val current = ++round
        engine.startListening(
            language,
            emptyList(),
            object : RecognizerListener {
                override fun onSpeechStart() = Unit

                override fun onPartial(text: String) {
                    if (current != round || !listening || text.isBlank()) return
                    heard = text.trim()
                    listener.onInterim(heard)
                }

                override fun onResult(text: String, confidence: Double?) {
                    if (current == round) settle(text.trim().ifEmpty { heard })
                }

                override fun onError(error: RecognizerError) {
                    if (current != round) return
                    if (heard.isNotEmpty()) settle(heard) else end { it.onProblem(problemOf(error)) }
                }
            },
        )
    }

    /** Ends the round now, keeping what was heard (nothing heard: [VoiceProblem.NO_SPEECH]). */
    fun stop() {
        if (!listening) return
        engine.cancel()
        settle(heard)
    }

    /** Abandons the round without reporting anything. */
    fun cancel() {
        if (!listening) return
        engine.cancel()
        listening = false
        listener = null
        round++
    }

    /** Releases the recognizer (the screen's view model is gone). */
    fun release() {
        cancel()
        engine.destroy()
    }

    private fun settle(text: String) {
        if (text.isBlank()) end { it.onProblem(VoiceProblem.NO_SPEECH) } else end { it.onFinal(text) }
    }

    private fun end(report: (Listener) -> Unit) {
        if (!listening) return
        val current = listener
        listening = false
        listener = null
        round++
        current?.let(report)
    }

    companion object {
        /** The language questions are recognised in (and answers read aloud in). */
        const val LANGUAGE = "es-ES"

        fun problemOf(error: RecognizerError): VoiceProblem = when (error) {
            RecognizerError.NO_SPEECH -> VoiceProblem.NO_SPEECH
            RecognizerError.PERMISSION_DENIED -> VoiceProblem.PERMISSION_DENIED
            RecognizerError.UNAVAILABLE -> VoiceProblem.UNAVAILABLE
            RecognizerError.BUSY, RecognizerError.TRANSIENT -> VoiceProblem.FAILED
        }
    }
}

File: web/src/tutor/voiceQuestion.ts
/**
 * One spoken question (#82): a single Web Speech recognition in Spanish, not continuous, with its
 * interim text, which ends when the student stops talking (or `stop()` is called) and hands over
 * the final text. The capture page's continuous transcriber (`capture/webSpeechTranscriber.ts`) is
 * the other use of the API; this one never talks to a session.
 */

import { speechRecognitionConstructor, WEB_SPEECH_LANGUAGE } from "../capture/webSpeechTranscriber";

export type VoiceProblemCode = "unsupported" | "permission-denied" | "no-speech" | "network" | "unavailable";

export interface VoiceQuestionCallbacks {
  /** The question so far, while the student speaks. */
  onInterim?: (text: string) => void;
  /** The recognized question; not called when nothing was understood. */
  onFinal: (text: string) => void;
  /** Why no question came out. */
  onProblem: (code: VoiceProblemCode) => void;
  /** The recognition is over, whatever the outcome. */
  onEnd?: () => void;
}

/** A running recognition: `stop()` ends it and keeps what was said so far. */
export interface VoiceQuestionHandle {
  stop(): void;
}

export type VoiceQuestionStarter = (callbacks: VoiceQuestionCallbacks) => VoiceQuestionHandle;

interface Alternative {
  readonly transcript: string;
}

interface Result extends ArrayLike<Alternative> {
  readonly isFinal: boolean;
}

interface Recognition {
  lang: string;
  continuous: boolean;
  interimResults: boolean;
  maxAlternatives: number;
  onresult: ((event: { results: ArrayLike<Result> }) => void) | null;
  onerror: ((event: { error: string }) => void) | null;
  onend: ((event: Event) => void) | null;
  start(): void;
  stop(): void;
}

const PROBLEMS: Record<string, VoiceProblemCode> = {
  "not-allowed": "permission-denied",
  "service-not-allowed": "permission-denied",
  "no-speech": "no-speech",
  network: "network",
};

export function voiceQuestionSupported(): boolean {
  return speechRecognitionConstructor() !== null;
}

/** Starts listening for one question (see the module comment). */
export const listenForQuestion: VoiceQuestionStarter = (callbacks) => {
  const Constructor = speechRecognitionConstructor();
  if (Constructor === null) {
    callbacks.onProblem("unsupported");
    callbacks.onEnd?.();
    return { stop: () => undefined };
  }
  const recognition = new Constructor() as unknown as Recognition;
  recognition.lang = WEB_SPEECH_LANGUAGE;
  recognition.continuous = false;
  recognition.interimResults = true;
  recognition.maxAlternatives = 1;
  let settled = "";
  let problem: VoiceProblemCode | null = null;
  let over = false;

  recognition.onresult = (event) => {
    const finals: string[] = [];
    const interims: string[] = [];
    for (let index = 0; index < event.results.length; index += 1) {
      const result = event.results[index];
      const text = result.length > 0 ? result[0].transcript.trim() : "";
      if (text === "") continue;
      (result.isFinal ? finals : interims).push(text);
    }
    settled = finals.join(" ");
    callbacks.onInterim?.([...finals, ...interims].join(" "));
  };
  recognition.onerror = (event) => {
    if (event.error === "aborted") return;
    problem ??= PROBLEMS[event.error] ?? "unavailable";
  };
  recognition.onend = () => {
    if (over) return;
    over = true;
    if (settled !== "") callbacks.onFinal(settled);
    else callbacks.onProblem(problem ?? "no-speech");
    callbacks.onEnd?.();
  };
  try {
    recognition.start();
  } catch {
    over = true;
    callbacks.onProblem("unavailable");
    callbacks.onEnd?.();
  }
  return {
    stop() {
      if (!over) recognition.stop();
    },
  };
};
